dali/pipeline/operator/builtin/external_source.h

```diff
@@ -123,20 +122,17 @@ class ExternalSource : public Operator<Backend> {
     return "ExternalSource (" + output_name_ + ")";
   }
 
-
-  /**
-   * @brief Sets the data that should be passed out of the op
-   * on the next iteration.
-   */
-  template<typename SrcBackend>
-  inline void SetDataSource(const TensorList<SrcBackend> &tl, cudaStream_t stream = 0) {
+  template<typename SrcBackend, template<typename> class SourceDataType>
+  inline void SetDataSourceHelper(const SourceDataType<SrcBackend> &t, cudaStream_t stream = 0) {
     if (std::is_same<SrcBackend, GPUBackend>::value && std::is_same<Backend, CPUBackend>::value) {
-      std::string msg = "Incorrect Backends warning. Loading GPU-originated data into CPU "
-                        "ExternalSource operator is discouraged and might be inefficient.";
-      DALI_WARN(msg);
+      DALI_WARN("Incorrect Backends warning. Loading GPU-originated data into CPU " +
```

Comment:
I have a feeling that we agreed that this will be an error instead of warning, yet it was merged this way.
Also, we will spam the user with the warning every iteration, won't we?

Reply:
And if we're just with the warning it probably should warn against the scenario of 
GPU input -> CPU external source -> GPU operator which is the main problem here.

Reply:
@szalpal ?
> CPU external source -> GPU operator 

It would require checking the graph, what is not very trivial now and beyond the scope of this PR.
I think this warning is good enough to let know the user that he should not do that, and if he get it every iteration it won't slow things down as it is already nit the most efficient approach anyway.

Reply:
We in fact agreed to have an error here. However, I went with warning, so DaliBackend development reasons - needed to verify GPU input there and I thought this way be easier than figuring out how to combine them with not-ready-yet python-level GPU input.

On the second thought for our previous discussion, I'd go for warning. There's no point of forbidding something just because it's tad slow and on the other hand it might come in handy when developing (e.g. there's GPU input, but not all GPU ops are ready, so we need partially CPU pipeline)